### ISSUE TITLE
ArrayIterator::append expects never

### ISSUE THREAD

**Issue by UrsusArctosUa** [vimeo/psalm#11117]
I'm not sure what 'never' means in the context of this error, but shouldn't this method expect 'mixed'?
https://psalm.dev/r/ef749b88b3
The code in the example should work fine with all versions since ArrayIterator was introduced.
https://3v4l.org/bUohU

**Comment by weirdan** [comment]
The types for the generic parameters (`TKey`, `TValue`) of ArrayIterator are inferred from the first argument of its constructor, it's being an empty array by default. An empty array's key and value types are most definitely not mixed. Psalm assumes them to be `never` (a bottom type in PHP). Therefore you're getting `ArrayIterator<never, never>`. As the result of maintaining this invariant, the parameter type of `append()` is also `never` - effectively you declared that you have an empty list and want it to remain this way. 

A language with built-in generics might require you to explicitly specify the generic types in cases like this, e.g. `let iterator = new ArrayIterator<int, mixed>();`

TL;DR - you can override Psalm's inference by adding a `@var` docblock - in fact it's one of the few constructs where you really _need_ it: https://psalm.dev/r/ce57a92807 . `@var` docblock serves as an explicit type specification here.

**Comment by UrsusArctosUa** [comment]
I see, thanks. I could never say that I understand templates.

**Comment by timoschinkel** [comment]
As an addendum, this is also how I managed to make this work for classes that extend `ArrayIterator`: https://psalm.dev/r/5d1eb309cc

It's not ideal, but it will do for now.
